### ISSUE TITLE
Improve gallery examples

### ISSUE THREAD

**Issue by jarrodmillman** [networkx/networkx#4261]
### Restructure

#4422 #4427 #4428 #4429 

### Existing examples
- [x] examples/drawing/plot_house_with_colors.py #4263 
- [x] examples/drawing/plot_four_grids.py #4264 
- [x] examples/drawing/plot_degree_rank.py #4259 (see also: #4265)
- [x] examples/drawing/plot_labels_and_colors.py #4266 
- [x] examples/drawing/plot_giant_component.py #4267 
- [x] examples/drawing/plot_knuth_miles.py #4251 
- [x] examples/drawing/plot_chess_masters.py #4252 
- [x] examples/basic/plot_read_write.py #4260 
- [ ] examples/graph/plot_napoleon_russian_campaign.py
- [x] examples/algorithms/plot_decomposition.py #4278 
- [ ] examples/algorithms/plot_krackhardt_centrality.py
- [x] examples/graph/plot_words.py #4409 
- [x] examples/applications/plot_circuits.py #4408 
- [x] examples/algorithms/plot_rcm.py #4411 

### New examples
- [x] gene-gene network #4269

### Geospatial examples
 - [x] pysal, geopandas, momepy, OSMnx, etc. #4366, #4381, #4383, #4407 

### External libraries examples

We should move the javascript, jit, and pygraphviz libraries here.  I think it makes sense to keep geospatial separate as it looks like it will have a number of specialized examples.

- [x] igraph #4404
- [ ] graphtools (maybe via igraph?)
- [ ] scikit-learn (maybe a spectral clustering example?)
- [ ] RAPIDS cuGraph
    - https://medium.com/rapids-ai/rapids-cugraph-networkx-compatibility-d119e417557c
    - https://github.com/ptheywood/cuda-cmake-github-actions
    - FLAMEGPU/FLAMEGPU2_dev#290
- [ ] graph plotting libraries ?

Are there other examples we should include?  I know a lot of libraries build on networkx.  It would be nice to have list of ones that provide functions to convert to/from NetworkX.

**Comment by eikNamo** [comment]
We would like to contribute a scikit-learn example to the gallery @SA-networkx
So far we are thinking of coming up with some clustering example or feature extraction from a graph.

**Comment by YaoM1886** [comment]
Hi, I and @eikNamo are working together on this issue. We have got this initial result by plotting spectral clustering in scikit-learn with networkx graph object. Hope this helps!
<img width="550" alt="scikit-learn-sc" src="https://user-images.githubusercontent.com/43605407/108960360-bf419500-76b0-11eb-910d-c28829d3a6d2.png">

**Comment by dschult** [comment]
It's not at all clear what is shown here.

**Comment by YaoM1886** [comment]
> It's not at all clear what is shown here.

Thanks for your reply:) The first subplot is a dataset composed of two categories.  The dataset is then used as input for spectral clustering(without labels). The second subplot is the graph visualization of the clustering result in which each node represents each data point and each edge represents similarity between two nodes by affinity matrix. 

Since spectral clustering algorithm is based on graph optimization, this example provides a chance for visualizing spectral clustering result in scikit-learn with the use of networkx graph object, so that users could manipulate the graph more easily to understand the clustering result.

**Comment by sharwinbobde** [comment]
Hello community!
I am working with @YaoM1886 on this and have done some more additions to our code to make this a more complete example/demonstration.

![image](https://user-images.githubusercontent.com/31354242/109389713-d0d3a880-790d-11eb-96c4-cc4ffce7e60f.png)

- You can see the true data and the clustering output in the feature space. This shows the readers that although items with features can be turned into a graph, the representation won't be perfect. (this honesty is necessary).
- The force-directed layout ensures the nodes will be similarly arranged with respect to the items in the feature space.
- The comparison of the three additionally shows the uninitiated readers (in Data Science) the true applications of such a method: representing items in a network and creating avenues to using graph algorithms on feature-space information.

**Comment by dschult** [comment]
Please label your axes. It is hard to determine what the scatterplots mean without some context.

**Comment by sharwinbobde** [comment]
> Please label your axes. It is hard to determine what the scatterplots mean without some context.

Ah yes, it is a toy example made using scikit-learn's `datasets.make_moons()`. We will label them as feature 0 and 1 and add a legend

**Comment by sharwinbobde** [comment]
So here is what we decided to do
## Task overview
- we do everything on a toy dataset
- we know data comes from 2 distributions only to compare results.
- the data is noisy (to show actual problems with clustering).
- we will use Spectral Clustering to (I) understand the data and (ii) cluster it
- we can also build a network using the affinity matrix which results in the following:
     - nodes are data points from the feature space
     - edges' weight is the value in the affinity matrix
     - nodes have cluster membership information
     
#### Usage
- this can be used to understand unlabelled data.
- The resulting graph is a condensed form of the feature space information
- we can indirectly use graph algorithms on feature space information
 

## Step 1
- Generate 2-class data which is not linearly separable 
```
n_samples = 100
X, y = datasets.make_moons(n_samples=n_samples)
```
- Add gaussian noise to make the problem difficult: `X = X + np.random.normal(1, 0.1, X.shape)`

## Step 2
- Lets conveniently select $N_C=2$ as the number of clusters. (this is just an example/demonstration :) )
- We will use *nearest-neighbour* based affinity matrix for [SpectralClustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.SpectralClustering.html). This is what the method does under the hood:
     - Given only the feature space information, find 10 nearest neighbours for each data point using euclidean distance.
     - create the affinity matrix. Here is the code snippet from the scikit-learn source code. 
     ```
     connectivity = kneighbors_graph(X, n_neighbors=self.n_neighbors,
                                            include_self=True,
                                            n_jobs=self.n_jobs)
     self.affinity_matrix_ = 0.5 * (connectivity + connectivity.T)
     ```
     - this is what a sample affinity sparse matrix looks like:
      ```
  (0, 76)	0.5
  (0, 70)	0.5
  (0, 61)	0.5
  (0, 36)	0.5
  (0, 66)	1.0
  (0, 92)	1.0
  (0, 15)	1.0
...
  (98, 83)	1.0
  (98, 99)	1.0
  (98, 98)	1.0
  (99, 25)	0.5
  (99, 20)	1.0
  (99, 48)	1.0
      ```
     - label the points according to the cluster

**Comment by sharwinbobde** [comment]
@dschult hope this is clearer :)
If there are any doubts we will keep updating the above explanation.
We do believe from our side that this is a strong direction for the Spectral Clustering example because it covers many possibilities of usage while still showing the limitations.
